src/functors.mli

```diff
@@ -0,0 +1,291 @@
+  : HETEROGENEOUS_MAP
+    with type 'a key = 'a Key.t
+      and type ('k,'m) value = ('k,'m) Value.t
+      and type 'm t = 'm Node.t
+
+(** Create an heterogeneous set with a custom {!NODE}.
+    @since v0.10.0 *)
+module MakeCustomHeterogeneousSet
+    (Key: HETEROGENEOUS_KEY)
+    (NODE: NODE with type 'a key = 'a Key.t and type ('key,'map) value = unit)
+  : HETEROGENEOUS_SET
+    with type 'a elt = 'a Key.t
+      and type 'a BaseMap.t = 'a NODE.t
+
+(** {1:hash_consed Hash-consed maps and sets} *)
+(** Hash-consed maps and sets uniquely number each of their nodes.
+    Upon creation, they check whether a similar node has been created before,
+    if so they return it, else they return a new node with a new number.
+    With this unique numbering:
+    - [equal] and [compare] become constant time operations;
+    - two maps with the same bindings (where keys are compared by {!KEY.to_int} and
+      values by {!HASHED_VALUE.polyeq}) will always be physically equal;
+    - functions that benefit from sharing, like {!BASE_MAP.idempotent_union} and
+      {!BASE_MAP.idempotent_inter} will see improved performance;
+    - constructors are slightly slower, as they now require a hash-table lookup;
+    - memory usage is increased: nodes store their tags inside themselves, and
+      a global hash-table of all built nodes must be maintained.
+      This is quickly amortized if multiple identical nodes are built,
+      as only one will be kept in memory.
+    - hash-consed maps assume their keys and values are immutable, where regular
+      maps can mutate values freely;
+    - {b WARNING:} when using physical equality as {!HASHED_VALUE.polyeq}, some
+      {b maps of different types may be given the same identifier}. See the end of
+      the documentation of {!HASHED_VALUE.polyeq} for details.
+      Note that this is the case in the default implementations {!HashedValue}
+      and {!HeterogeneousHashedValue}.
+
+    All hash-consing functors are {b generative}, since each functor call will
+    create a new hash-table to store the created nodes. Calling a functor
+    twice with same arguments will lead to two numbering systems for identifiers,
+    and thus the types should not be considered compatible.  *)
+
+(** Hash-consed version of {!MAP}. See {!hash_consed} for the differences between
+    hash-consed and non hash-consed maps.
+
+    This is a generative functor, as calling it creates a new hash-table to store
+    the created nodes, and a reference to store the next unallocated identifier.
+    Maps/sets from different hash-consing functors (even if these functors have
+    the same arguments) will have different (incompatible) numbering systems and
+    be stored in different hash-tables (thus they will never be physically equal).
+
+    @since v0.10.0 *)
+module MakeHashconsedMap(Key: KEY)(Value: HASHED_VALUE)() : sig
+  include MAP_WITH_VALUE with type key = Key.t and type 'a value = 'a Value.t (** @closed *)
+
```

Comment:
It seems to me that you could share the to_int/equal/compare triple in all these signatures, and some of their comments

Reply:
Done